contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapV2Reserve.sol

```diff
@@ -0,0 +1,324 @@
+     */
+    function trade(
+        IERC20 srcToken,
+        uint256 srcAmount,
+        IERC20 destToken,
+        address payable destAddress,
+        uint256 conversionRate,
+        bool /* validate */
+    ) external override payable returns (bool) {
+        require(tradeEnabled, "trade is disabled");
+        require(msg.sender == kyberNetwork, "only kyberNetwork");
+        require(isValidTokens(srcToken, destToken), "token is not listed");
+
+        require(conversionRate > 0, "conversionRate 0");
+        if (srcToken == ETH_TOKEN_ADDRESS) {
+            require(msg.value == srcAmount, "msg.value != srcAmount");
+        } else {
+            require(msg.value == 0, "msg.value is not 0");
+        }
+
+        uint256 expectedDestAmount = calcDestAmount(
+            srcToken,
+            destToken,
+            srcAmount,
+            conversionRate
+        );
+
+        uint256 destAmount;
+        (uint256 actualRate, address[] memory path) = calcUniswapConversion(
+            srcToken,
+            destToken,
+            srcAmount
+        );
+        require(conversionRate <= actualRate, "expected conversionRate <= actualRate");
+
+        if (srcToken == ETH_TOKEN_ADDRESS) {
+            // Deduct fees (in ETH) before converting
+            uint256 quantity = deductFee(srcAmount);
+
+            uint256[] memory amounts = uniswapRouter.swapExactETHForTokens{value: quantity}(
+                expectedDestAmount,
+                path,
+                destAddress,
+                DEADLINE
+            );
+            destAmount = amounts[amounts.length - 1];
+            require(destAmount >= expectedDestAmount, "Returned trade amount too low");
+        } else {
+            srcToken.safeTransferFrom(msg.sender, address(this), srcAmount);
+            uint256[] memory amounts = uniswapRouter.swapExactTokensForETH(
+                srcAmount,
+                expectedDestAmount,
+                path,
+                address(this),
+                DEADLINE
+            );
+
+            destAmount = amounts[amounts.length - 1];
+            // Deduct fees (in ETH) after converting
+            destAmount = deductFee(destAmount);
+            require(destAmount >= expectedDestAmount, "Returned trade amount too low");
+            // Transfer user-expected dest amount
+            destAddress.transfer(expectedDestAmount);
+        }
+
+        emit TradeExecute(
+            msg.sender,
+            srcToken,
+            srcAmount,
+            destToken,
+            expectedDestAmount,
+            destAddress
+        );
+        return true;
+    }
+
+    function setFee(uint256 _feeBps) external onlyAdmin {
+        require(_feeBps < BPS, "fee >= BPS");
+        if (_feeBps != feeBps) {
+            feeBps = _feeBps;
+            emit FeeUpdated(_feeBps);
+        }
+    }
+
+    function listToken(
+        IERC20 token,
+        bool addDefaultPaths,
+        bool validate
```

Comment:
thinking again.

addDefaultPaths should be enough.

if it fails.
just call again with same values

Lets discuss.